### PR TITLE
Pick up openstack info from secret for cinder

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:
Enhance cinder code to pick up a secret with keystone authentication info so we can allow folks to use their own projects (as opposed to the single admin specified when kubernetes control plane is started)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by edisonxiang** [comment]
@dims This is a very good feature for openstack provider:)
I pulled this pr and tested it by ```local-up-cluster``` and set ```/etc/cloud.conf```
but did not set the secret in kube-system.
I got a error like below and the node can not be added into the k8s master.
```Kubelet failed to get node info: failed to get external ID from cloud provider: instance not found```
```kubectl get nodes``` returned ```No resources found.```
Also I checkout master and rebuild it, it is ok to get nodes.
https://gist.github.com/edisonxiang/bf7e1d74b0ec201ea79407d4f6b51a0d

**Comment by dims** [comment]
@edisonxiang : is `ecs-edison` the instance name of the VM as known to OpenStack?

**Comment by edisonxiang** [comment]
> @edisonxiang : is ecs-edison the instance name of the VM as known to OpenStack?
@dims yep

**Comment by edisonxiang** [comment]
@dims I rebase your another pr with this one.
 (https://github.com/kubernetes/kubernetes/pull/61890)
This error is gone:)
So the previous error is not related with your current pr.
lgtm, Thanks.

**Comment by dims** [comment]
Great! thanks @edisonxiang i'll rebase this again today

**Comment by liggitt** [comment]
@kubernetes/sig-storage-api-reviews

**Comment by liggitt** [comment]
Left some comments on https://github.com/kubernetes/kubernetes/commit/a83c62cc4ad7733f3a1c7ddad6f85b5b6c792faf in case they get orphaned

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -723,11 +723,16 @@ type CinderVolumeSource struct {
...
+	// sensitive information to connect to openstack. This may be
+	// empty if no secret object is specified.
+	// +optional
+	SecretRef *SecretReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`
 	// Optional: Defaults to false (read/write). ReadOnly here will force
 	// the ReadOnly setting in VolumeMounts.
 	// More info: https://releases.k8s.io/HEAD/examples/mysql-cinder-pd/README.md
 	// +optional
-	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,3,opt,name=readOnly"`
+	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,4,opt,name=readOnly"`
```
[the remark this answers is not in the document]
Comment (dims) [reply staging/src/k8s.io/api/core/v1/types.go]: Ack. fixing

File: pkg/printers/internalversion/describe.go [hunk 2]
```diff
@@ -966,6 +966,14 @@ func printCinderVolumeSource(cinder *api.CinderVolumeSource, w PrefixWriter) {
 		cinder.VolumeID, cinder.FSType, cinder.ReadOnly)
 }
 
+func printCinderPersistentVolumeSource(cinder *api.CinderPersistentVolumeSource, w PrefixWriter) {
+	w.Write(LEVEL_2, "Type:\tCinder (a Persistent Disk resource in OpenStack)\n"+
```
Comment (rootfs) [inline pkg/printers/internalversion/describe.go]: no secretRef?

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -110,18 +110,6 @@ START_MODE=${START_MODE:-"all"}
 # A list of controllers to enable
 KUBE_CONTROLLERS="${KUBE_CONTROLLERS:-"*"}"
 
-# sanity check for OpenStack provider
-if [ "${CLOUD_PROVIDER}" == "openstack" ]; then
-    if [ "${CLOUD_CONFIG}" == "" ]; then
```
Comment (rootfs) [inline hack/local-up-cluster.sh]: you can use default config here

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 4]
```diff
@@ -225,37 +230,98 @@ func configFromEnv() (cfg Config, ok bool) {
...
 	cfg.Metadata.DHCPDomain = "novalocal"
 	cfg.BlockStorage.BSVersion = "auto"
-
-	return
 }
 
-func readConfig(config io.Reader) (Config, error) {
-	if config == nil {
-		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
+func configFromSecret(cfg *Config, kubeClient clientset.Interface, name string, namespace string) error {
```
Comment (rootfs) [inline pkg/cloudprovider/providers/openstack/openstack.go]: I suggest just pass secret here not to get from api server. This saves a rbac rule for openstack cloud provider.

File: pkg/volume/cinder/cinder.go [hunk 5]
```diff
@@ -193,31 +235,12 @@ func (plugin *cinderPlugin) newProvisionerInternal(options volume.VolumeOptions,
...
-		}
-	}
-
-	switch cloud := cloud.(type) {
-	case *openstack.OpenStack:
-		return cloud, nil
-	default:
-		return nil, errors.New("invalid cloud provider: expected OpenStack")
+func (plugin *cinderPlugin) getCloudProvider(name string, namespace string) (BlockStorageProvider, error) {
+	cloud, err := openstack.NewOpenStack(plugin.host.GetKubeClient(), name, namespace)
```
Comment (rootfs) [inline pkg/volume/cinder/cinder.go]: get secret here and pass into `NewOpenStack`

File: pkg/volume/cinder/cinder.go [hunk 6]
```diff
@@ -120,27 +115,66 @@ func (plugin *cinderPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
...
+			return localSecretRef.Name, defaultNamespace, nil
+		}
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.Cinder != nil {
+		secretRef := spec.PersistentVolume.Spec.Cinder.SecretRef
+		if secretRef != nil {
+			return secretRef.Name, secretRef.Namespace, nil
+		}
+	} else {
+		return "", "", fmt.Errorf("Spec does not reference an Cinder volume type")
```
Comment (rootfs) [inline pkg/volume/cinder/cinder.go]: an -> a

File: pkg/apis/core/types.go [hunk 7]
```diff
@@ -988,6 +988,30 @@ type CinderVolumeSource struct {
 	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
 	// +optional
 	FSType string
+	// Optional: SecretRef is reference to the connection information for openstack, default is empty.
+	// +optional
+	SecretRef *LocalObjectReference
```
Comment (rootfs) [inline pkg/apis/core/types.go]: the comment is different from that in v1/types.go

File: pkg/printers/internalversion/describe.go (hunk 2 above)
Comment (dims) [reply pkg/printers/internalversion/describe.go]: Fixing.

File: hack/local-up-cluster.sh (hunk 3 above)
Comment (dims) [reply hack/local-up-cluster.sh]: Right, we don't need to do extra validation for just OpenStack.

File: pkg/apis/core/types.go (hunk 7 above)
Comment (dims) [reply pkg/apis/core/types.go]: Fixing

**Comment by dims** [comment]
@liggitt the basic information is still in the cloud config. new behavior is to allow folks can override to use specific tenant for their cinder volume using secrets.

File: pkg/volume/cinder/cinder.go (hunk 6 above)
Comment (dims) [reply pkg/volume/cinder/cinder.go]: fixed

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 4 above)
Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: @rootfs that will help just the cinder scenario. the code is structured so we can customize LB(s) next.

File: pkg/volume/cinder/cinder.go (hunk 5 above)
Comment (dims) [reply pkg/volume/cinder/cinder.go]: Please see response in the comment on configFromSecret

File: staging/src/k8s.io/api/core/v1/types.go [hunk 8]
```diff
@@ -728,6 +728,43 @@ type CinderVolumeSource struct {
...
 	// +optional
 	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,3,opt,name=readOnly"`
+	// Optional: points to a secret object that has enough information to connect
+	// to OpenStack. The secret data can contain fields from the "[Global]" section
+	// of the cloud configuration file. The values of these fields essentially
+	// override the values from the main configuration file set up by the
+	// administrator. This is helpful for example to customize the "domain-id",
+	// "tenant-id", "user-id" and "password" for specific pod(s)
+	// +optional
+	SecretRef *LocalObjectReference `json:"secretRef,omitempty" protobuf:"bytes,4,opt,name=secretRef"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go]: keep the struct order the same between internal and external types (..., FSType, ReadOnly, SecretRef)

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 9]
```diff
@@ -225,37 +230,98 @@ func configFromEnv() (cfg Config, ok bool) {
...
+	if name == "" {
+		name = "openstack"
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: I wouldn't recommend defaulting to a secret name/namespace like this... is this new behavior?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 10]
```diff
@@ -225,37 +230,98 @@ func configFromEnv() (cfg Config, ok bool) {
...
+			case "trust-id":
+				cfg.Global.TrustID = value
+			case "domain-id":
+				cfg.Global.DomainID = value
+			case "domain-name":
+				cfg.Global.DomainName = value
+			case "region":
+				cfg.Global.Region = value
+			case "ca-file":
+				cfg.Global.CAFile = value
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: does it make sense to reference a filepath on the apiserver machine via a config value loaded from an API resource like a secret?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 11]
```diff
@@ -225,37 +230,98 @@ func configFromEnv() (cfg Config, ok bool) {
...
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		for name, data := range secret.Data {
+			value := string(data)
+			switch name {
+			case "auth-url":
+				cfg.Global.AuthURL = value
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: partial overrides of the config like this lets you do things like override the URL that the credentials loaded from the cloud provider config file are sent to... that doesn't seem like something we'd want

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 4 above)
Comment (liggitt) [reply pkg/cloudprovider/providers/openstack/openstack.go]: I tend to agree that the config should be fetched outside and passed in (from file, from secret, etc). That makes it clearer what components are interacting with the kube API, since cloud provider code is exercised from multiple components

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 12]
```diff
@@ -225,37 +230,98 @@ func configFromEnv() (cfg Config, ok bool) {
...
 
-	cfg, _ := configFromEnv()
+	// Load what we can find in the environment variables
+	configFromEnv(&cfg)
 
 	// Set default values for config params
 	cfg.BlockStorage.BSVersion = "auto"
 	cfg.BlockStorage.TrustDevicePath = false
-	cfg.BlockStorage.IgnoreVolumeAZ = false
+	cfg.BlockStorage.IgnoreVolumeAZ = true
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: is this related to this PR?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 13]
```diff
@@ -313,6 +379,34 @@ func checkOpenStackOpts(openstackOpts *OpenStack) error {
 	return checkMetadataSearchOrder(openstackOpts.metadataOpts.SearchOrder)
 }
 
+// NewOpenStack can be used when we need to use a specific secret
+func NewOpenStack(kubeClient clientset.Interface, name string, namespace string) (*OpenStack, error) {
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: I'm trying to understand the implications of having lots of instances of the openstack cloud provider, each talking to potentially different openstack backends, or as different tenants, etc. Is there precedent for constructing per-operation cloud provider instances like this? We need to make sure all the components that deal with cloud providers have the correct instance (this PR opens the possibility means they could have an inconsistently constructed instances)

At first glance, it seems like it might be safer to make the specific operations (volume operations, LB operations) that you want sharded by tenant take additional input (like the new config field in the Cinder volume source), rather than using completely different cloud provider instances.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 8 above)
Comment (dims) [reply staging/src/k8s.io/api/core/v1/types.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 9 above)
Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: ack. nuking, we can add it in another PR if required

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 10 above)
Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Ack removing ca-file and auth-url

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 11 above)
Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Ack removing ca-file and auth-url

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 12 above)
Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Sorry. this sneaked in. fixing

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 13 above)
Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: i've removed the auth-url and ca-file, so we will talk only to the openstack configured by the administrator. 

So we are left with being able to configure specific tenant/domain/user/password. which is fine for both load balancers and volumes.

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 14]
```diff
@@ -225,27 +230,50 @@ func configFromEnv() (cfg Config, ok bool) {
...
-	if config == nil {
-		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
+func configFromSecret(cfg *Config, kubeClient clientset.Interface, name string, namespace string) error {
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		for name, data := range secret.Data {
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: I don't know enough about these options to know if it's reasonable to mix these with the global config... if this override is provided, should we start by emptying out these values from the global config? The scenarios I'm thinking of most are:
* mutually-exclusive options (e.g. global config specifies tenant-id, local config specifies tenant-name). I assume there's a precedence order the library uses those in, and it seems weird to require the local config to null out the global value to ensure it's value takes effect
* reuse of globally provided credentials. I can't tell if this is a feature or a security hole, but as is, this would let you use globally configured credentials against a different tenant/domain/region, right?

Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: An the other point is, the global config has the admin credentials, we are really trying to help end users use their own credentials (and their own tenant). So in effect we are trying to get them to limit themselves ...

Comment (liggitt) [reply pkg/cloudprovider/providers/openstack/openstack.go]: > An the other point is, the global config has the admin credentials, we are really trying to help end users use their own credentials (and their own tenant). So in effect we are trying to get them to limit themselves ...

it has the global credentials, but only pointed at the tenant and region the cluster admin selected. I'm not sure we can assume a random pod author should be able to make use of those credentials to choose a different tenant or region, right?

Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Good point, if they specify a tenant or region, then i will make sure they should specify a different userid/name (and hence the associated domain and password)

**Comment by dims** [comment]
/hold

(for testing)

**Comment by dims** [comment]
testing with https://gist.github.com/dims/d3be058f90fecec8a48c8f8e295cef71 and https://gist.github.com/dims/fab4505551fadd9ec06675c5dcd44b28

File: pkg/volume/cinder/cinder.go (hunk 5 above)
Comment (liggitt) [reply pkg/volume/cinder/cinder.go]: reconstructing the cloud provider here is a very different flow than any other plugin, and seems problematic for a few reasons:
* this is called frequently, and assumed to be low-cost (for example, constructing a client once per process, not once per cloudprovider access)
* it means the volume plugin cloudprovider is different than the controllermanager cloud provider, and has a different view of cloud provider resources. is that likely to cause problems doing things like finding instances, etc?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 15]
```diff
@@ -225,27 +230,75 @@ func configFromEnv() (cfg Config, ok bool) {
...
+func configFromSecret(cfg *Config, kubeClient clientset.Interface, name string, namespace string) error {
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		// When a secret is specified with some of the keys below, we should make sure it has a specific
+		// user id/name and should not use the global admin credentials.
+		specialKeys := []string{"tenant-id", "tenant-name", "region"}
```
Comment (liggitt) [inline pkg/cloudprovider/providers/openstack/openstack.go]: sorry to keep pushing on this, but this overlaying still seems problematic. why wouldn't we require them to provide credentials if they want to point to different things than the cluster-admin configured? some currently allowed examples that seem suspicious:
* they can swap a different user-id/username and keep the cluster-admin specified password
* they can swap a different domain-id/domain-name and keep the cluster-admin specified username/password

**Comment by dims** [comment]
/hold

after offline chat with @liggitt .. i'll rework this to just the API change. will leave the implementation for later, may be in the external cloud provider.

**Comment by dims** [comment]
Moving just the API changes here to another PR - https://github.com/kubernetes/kubernetes/pull/63826

Thanks,
Dims

/close
